sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpProducer.cs

```diff
@@ -454,23 +470,17 @@ public override async Task CloseAsync(CancellationToken cancellationToken)
 
                 while (!cancellationToken.IsCancellationRequested)
                 {
+                    EventHubsEventSource.Log.EventPublishStart(EventHubName, logPartition, operationId);
+
                     try
                     {
-                        using AmqpMessage batchMessage = messageFactory();
-
-                        // Creation of the link happens without explicit knowledge of the cancellation token
-                        // used for this operation; validate the token state before attempting link creation and
-                        // again after the operation completes to provide best efforts in respecting it.
-
-                        EventHubsEventSource.Log.EventPublishStart(EventHubName, logPartition, operationId);
+                        using AmqpMessage batchMessage = MessageConverter.CreateBatchFromMessages(messages);
```

Comment:
This approach requires converting each AMQP message into an `ArraySegment<byte>` to be packaged in the envelope.  For retry scenarios, this work is performed for each attempt.  The alternative would be to create the batch envelope once and then clone it for each send attempt.   That requires the first send attempt to pay for creation of the batch envelope creation and a clone, while retries pay only for the clone.

I chose to optimize for the golden path here - since a single send attempt is the majority outcome, by creating a new batch envelope during each attempt, the first send avoids paying for an unneeded clone.